operator/rpc_client.go

```diff
@@ -42,82 +75,172 @@ func (c *AggregatorRpcClient) dialAggregatorRpcClient() error {
 	return nil
 }
 
-func (c *AggregatorRpcClient) sendRequest(sendCb func() error, maxRetries int, retryInterval time.Duration) {
-	if c.rpcClient == nil {
-		c.logger.Info("rpc client is nil. Dialing aggregator rpc client")
-		err := c.dialAggregatorRpcClient()
+func (c *AggregatorRpcClient) InitializeClientIfNotExist() error {
+	c.rpcClientMutex.Lock()
+	defer c.rpcClientMutex.Unlock()
+
+	if c.rpcClient != nil {
+		return nil
+	}
+
+	return c.dialAggregatorRpcClient()
+}
+func (c *AggregatorRpcClient) onTick() {
+	tickerC := c.resendTicker.C
+	for {
+		// TODO(edwin): handle closed chan
+		<-tickerC
+
+		// Critically ugly section
+		{
+			c.unsentMessagesMutex.Lock()
+			if len(c.unsentMessages) == 0 {
+				c.unsentMessagesMutex.Unlock()
+				continue
+			}
+			c.unsentMessagesMutex.Unlock()
+		}
+
+		err := c.InitializeClientIfNotExist()
 		if err != nil {
-			c.logger.Error("Could not dial aggregator rpc client. Not sending signed task response header to aggregator. Is aggregator running?", "err", err)
-			return
+			continue
 		}
+
+		c.tryResendFromDeque()
+	}
+}
+
+// Expected to be called with initialized client.
+func (c *AggregatorRpcClient) tryResendFromDeque() {
+	c.unsentMessagesMutex.Lock()
+	defer c.unsentMessagesMutex.Unlock()
+
+	if len(c.unsentMessages) != 0 {
+		c.logger.Info("Resending messages from queue")
 	}
 
-	c.logger.Info("Sending data to aggregator")
-	for i := 0; i < maxRetries; i++ {
-		err := sendCb()
+	errorPos := 0
+	for i := 0; i < len(c.unsentMessages); i++ {
+		message := c.unsentMessages[i]
+
+		// Assumes client exists
+		var err error
+		var reply bool
+
+		switch message.MessageType {
+		case CheckpointType:
+			signedCheckpointTaskResponse := message.Message.(*coretypes.SignedCheckpointTaskResponse)
+			// TODO(edwin): handle error
+			err = c.rpcClient.Call("Aggregator.ProcessSignedCheckpointTaskResponse", signedCheckpointTaskResponse, &reply)
 
-		if err == nil {
-			c.logger.Info("Data successfully sent to aggregator")
-			return
+		case StateRootType:
+			signedStateRootUpdateMessage := message.Message.(*coretypes.SignedStateRootUpdateMessage)
+			err = c.rpcClient.Call("Aggregator.ProcessSignedStateRootUpdateMessage", signedStateRootUpdateMessage, &reply)
+
+		case OperatorSetUpdateType:
+			signedOperatorSetUpdateMessage := message.Message.(*coretypes.SignedOperatorSetUpdateMessage)
+			err = c.rpcClient.Call("Aggregator.ProcessSignedOperatorSetUpdateMessage", signedOperatorSetUpdateMessage, &reply)
+
+		default:
+			panic("unreachable")
 		}
 
-		c.logger.Infof("Retrying in %s", retryInterval.String())
-		time.Sleep(retryInterval)
+		if err != nil {
+			c.logger.Error("Couldn't resend message", "err", err)
+
+			c.unsentMessages[errorPos] = message
+			errorPos++
+		}
+	}
+
+	c.unsentMessages = c.unsentMessages[:errorPos]
```

Comment:
I wonder if having a separate array for messages that couldn't be re-sent, which would be simpler, wouldn't even be better perf-wise. We're still resizing the original array here, which might trigger a realloc anyway, and range loops are usually more efficient. Not blocking, but I had to stop for some minutes here to make sure this was safe, so maybe it's a worthwhile change.

Reply:
Tbh I don't mind 2nd array approach. Tho I would like to point out that the critical moment here is memory. In worst case there will be x2 of memory that array uses. With db we want to have a limit_size at which we going to dump them. So need to be careful here. 
Regarding reallocation I'm 99% sure it shalln't happen since we only decrease this of array - changing the header. So I will be very surprised if any reallocation happens. This is [standard algorithm ](https://en.cppreference.com/w/cpp/algorithm/remove)to avoid exactly that case.